Add option to map (force) port use on the server for reverse tunnels

siepkes opened the pull request:
This change adds a `port_mapping` option to the `ReverseTunnel` definition in the (YAML) restriction file.

It maps ports on the server side from X to Y (X:Y). Where X is the originally requested port by the client and Y is the port which will be used to listen on server-side.

For example with `10001:8080` configured and a client which connects using `-R tcp://10001:localhost:80` the server will listen on port 8080 instead of 10001. The originally requested ports (NOT the mapped ports) still needs to be allowed via the `ports` directive.

This is for example useful when dealing with lots of clients and you don't want to coordinate port use on all the clients but centrally on the server.

Let me know if this works for you and the project or if you have (other) ideas on it!

erebe commented on src/tunnel/server.rs:
```diff
@@ -94,7 +95,22 @@ async fn run_tunnel(
+                .iter()
+                .find_map(|allow| {
+                    if let AllowConfig::ReverseTunnel(allow) = allow {
+                        if remote.protocol.is_reverse_tunnel() {
```
No need of this check, as you already know at this point that a ReverseTcp has been requested.

siepkes replied:
Done!

erebe commented on src/tunnel/server.rs:
```diff
@@ -94,7 +95,22 @@ async fn run_tunnel(
+                .find_map(|allow| {
+                    if let AllowConfig::ReverseTunnel(allow) = allow {
+                        if remote.protocol.is_reverse_tunnel() {
+                            return allow.port_mapping.get(&remote.port).cloned();
```
.clone() is enough i would say

erebe commented:
Hello,

I am not against the idea of it, but it should be supported for all reverse tunnels :)
The PR only introduce it for ReverseTcp at the moment.

siepkes commented:
> I am not against the idea of it, but it should be supported for all reverse tunnels :)

Yes, definitely! Sorry, should have been clear about that it was merely a first proof on concept to iterate on and to get a feel about how you thought about such a feature. Since your open to it I've also implemented the other ones but I still need to test the thing (so hold yer horses with any merging :wink: ) 

> .clone() is enough i would say

I'm still new to Rust  and I'm struggling a bit with this one. Simply changing to `.clone() ` does not work since `cloned()` returns a borrowed value. I've also moved the logic to a separate method so that probably also compounds a bit to my lifetime issues. I assume the advantage in this context of `clone()`  vs `cloned()` is about preventing adding a new integer on the stack?
